service/dap/server.go

```diff
@@ -987,16 +1004,81 @@ func (s *Server) convertVariable(v *proc.Variable) (value string, variablesRefer
 			value = "<" + typeName + ">"
 		}
 		if len(v.Children) > 0 {
-			variablesReference = s.variableHandles.create(v)
+			variablesReference = maybeCreateVariableHandle(v)
 		}
 	}
 	return
 }
 
-// onEvaluateRequest sends a not-yet-implemented error response.
+// onEvaluateRequest handles 'evalute' requests.
 // This is a mandatory request to support.
-func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) { // TODO V0
-	s.sendNotYetImplementedErrorResponse(request.Request)
+// Support the following expressions:
+// -- {expression} - evaluates the expression and returns the result as a variable
+// -- call {function} - injects a function call and returns the result as a variable
+// TODO(polina): users have complained about having to click to expand multi-level
+// variables, so consider also adding the following:
+// -- print {expression} - return the result as a string like from dlv cli
+func (s *Server) onEvaluateRequest(request *dap.EvaluateRequest) {
+	showErrorToUser := request.Arguments.Context != "watch"
+	if s.debugger == nil {
+		s.sendErrorResponseWithOpts(request.Request, UnableToEvaluateExpression, "Unable to evaluate expression", "debugger is nil", showErrorToUser)
+		return
+	}
+	// Default to the topmost stack frame of the current goroutine in case
+	// no frame is specified (e.g. when stopped on entry or no call stack frame is expanded)
+	goid, frame := -1, 0
+	if sf, ok := s.stackFrameHandles.get(request.Arguments.FrameId); ok {
+		goid = sf.(stackFrame).goroutineID
+		frame = sf.(stackFrame).frameIndex
+	}
+	// TODO(polina): Support config settings via launch/attach args vs auto-loading?
+	apiCfg := &api.LoadConfig{FollowPointers: true, MaxVariableRecurse: 1, MaxStringLen: 64, MaxArrayValues: 64, MaxStructFields: -1}
+	prcCfg := proc.LoadConfig{FollowPointers: true, MaxVariableRecurse: 1, MaxStringLen: 64, MaxArrayValues: 64, MaxStructFields: -1}
+
+	response := &dap.EvaluateResponse{Response: *newResponse(request.Request)}
+	isCall, err := regexp.MatchString(`^\s*call\s+\S+`, request.Arguments.Expression)
+	if err == nil && isCall { // call {expression}
+		state, err := s.debugger.Command(&api.DebuggerCommand{
+			Name:                 api.Call,
+			ReturnInfoLoadConfig: apiCfg,
+			Expr:                 strings.Replace(request.Arguments.Expression, "call ", "", 1),
+			UnsafeCall:           false,
+			GoroutineID:          goid,
+		})
+		if _, isexited := err.(proc.ErrProcessExited); isexited || err == nil && state.Exited {
+			e := &dap.TerminatedEvent{Event: *newEvent("terminated")}
+			s.send(e)
+			return
+		}
+		if err != nil {
+			s.sendErrorResponseWithOpts(request.Request, UnableToEvaluateExpression, "Unable to evaluate expression", err.Error(), showErrorToUser)
+			return
+		}
+		retVars := s.debugger.ReturnValues(&prcCfg)
```

Comment:
The problem with this is that there's no guarantee that the return values for the call will be available the very next time the debugger stops. It could be that the debugger stopped because it encountered a breakpoint, either inside the function that was called or some other breakpoint in a different goroutine.

I'm not sure what the right way to handle this would be, I think it depends on what the UI of VSCode can do. For example what happens when using VSCode with gdb, doing a function call while a breakpoint is set in the function being called?

Reply:
I added a draft implementation. See TODO please - I am not fully sure what the cleanest way is to detect if the call completed or not.

Also, I added a unittest for a breakpoint inside of the call. I still need to add a test where the call restarts other goroutines that stop on a breakpoint. In the meantime, I tested this manually with https://github.com/aarzilli/delve_client_testing/blob/master/05nextbp/main.go. It works when one-off calls are issued from the debug console. But if we put a call into a watch panel, then everything gets totally screwed. The call restarts everything, so before it finishes, another goroutine stops. That triggers watch re-eval, but relative to another goroutine, so it is allowed. This again restarts everything and just keeps going like continue, never really stopping at any breakpoints for longer than a split second. Should we be disallowing calls as watch expressions?   That would be  unfortunate because it could be such a neat way to pretty print most interesting variables.

Reply:
> See TODO please - I am not fully sure what the cleanest way is to detect if the call completed or not.

You should iterate over the thread list and check if one has the goroutine ID you want and `ReturnValues != nil`. I don't think the way you are doing this works. First off there's no guarantee that the injected goroutine will make progress before we are interrupted by a simultaneous breakpoint, secondly consider what happens if the goroutine stopped on a breakpoint inside `main.f` and the user injects a call to `main.g` which calls `main.f` again.

> Should we be disallowing calls as watch expressions to avoid such circular behavior? That would be unfortunate because it could be such a neat way to pretty print most interesting variables in the watch panel and have them automatically update.

IMHO there's no way to do it cleanly. I'd like to know what VSCode's solution is for C-like languages, maybe there's something I haven't thought of.